File: AlgorithmStudy/src/y2022/jul/Solution0717.java
package y2022.jul;

public class Solution0717 {
    int mod = (int) (1e9) + 7;

    public int kInversePairs(int n, int k) {
        // dp

        /*
        1 ~ n 까지 중 k개 뒤집어져 있다.

         dp[n][k] = n개의 길이에서 k개 뒤집어진 수

         dp[0][0] = 0
         dp[1][0] = 1
         dp[1][1] = 0
         dp[2][1] = 1
             0   1   2   3   4   5
         0  [  ][  ][  ][  ][  ][  ]
         1  [ 1][  ][  ][  ][  ][  ]
         2  [ 1][ 1][  ][  ][  ][  ]
              ^   ^   ^
                      l -> 3개의 합이 [3][2]의 dp 값
         3  [ 1][ 2][ 2][ 1][  ][  ]
         4  [  ][  ][  ][  ][  ][  ]
         */

        int[][] dp = new int[n + 1][k + 1];
        for (int i = 1; i <= n; i++) {
            for (int j = 0; j <= k && j <= i * (i - 1) / 2; j++) {
                if (i == 1 && j == 0) {
                    dp[i][j] = 1;
                    break;
                } else if (j == 0) {
                    dp[i][j] = 1;
                }
                else {
                    int val = (dp[i - 1][j] + mod - ((j - i) >= 0 ? dp[i - 1][j - i] : 0)) % mod;
                    dp[i][j] = (dp[i][j - 1] + val) % mod;
                }
            }
        }
        return dp[n][k];


    }
}

File: AlgorithmStudy/src/y2023/jul/Solution0717.kt
package y2023.jul

import model.ListNode
import java.util.*

// https://leetcode.com/problems/add-two-numbers-ii/
class Solution0717 {
    fun addTwoNumbers(l1: ListNode?, l2: ListNode?): ListNode? {
        /*
        StringBuilder -> Long.valueOf() -> X

        list 를 뒤집은 다음에
        더해 나가는거
         */

        val stack1 = Stack<Int>()
        val stack2 = Stack<Int>()

        var l1 = l1
        var l2 = l2
        while (l1 != null) {
            stack1.push(l1.`val`)
            l1 = l1.next
        }

        while (l2 != null) {
            stack2.add(l2.`val`)
            l2 = l2.next
        }


        val stack3 = Stack<Int>()
        var carry = 0
        while (stack1.isNotEmpty() || stack2.isNotEmpty()) {
            val l = if (stack1.isNotEmpty()) stack1.pop() else 0
            val r = if (stack2.isNotEmpty()) stack2.pop() else 0

            val sum = l + r + carry
            carry = (sum / 10).toInt()

            stack3.push(sum % 10)
        }
        
        if (carry != 0) {
            stack3.push(carry)
        }
        
        val root = ListNode()
        var temp = root
        while (stack3.isNotEmpty()) {
            temp.next = ListNode(stack3.pop())
            temp = temp.next
        }
        
        return root.next
    }
}
